[Woo Jia Hao] iP

woojiahao opened the pull request:
# Cyrus

> Worry no more about your tasks!

**Cyrus** is a hassle free task list manager built with [Java.](https://www.java.com/en/) 

## Core features

* Add events
* Add deadlines
* Add regular todos
* Mark/unmark tasks
* Delete tasks
* Save tasks to file

## Building

To get started with `Cyrus`, you can use the following steps:

1. Clone the repository

```bash
git clone https://github.com/woojiahao/ip
```

2. Change to corresponding directory

```bash
cd ip/
```

3. Build the project

```bash
./gradlew shadowJar
```

4. Run the built jar

```bash
java -jar build/libs/duke.jar
```

## Pending features

- [ ] GUI for interaction

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

woojiahao committed:
Rename Duke to Cyrus and add greeting

woojiahao committed:
Setup echo of commands

woojiahao committed:
Remove unneeded semicolon

woojiahao committed:
Implement input storage and listing

woojiahao committed:
Implement marked status for tasks

woojiahao committed:
Set Task name to final

woojiahao committed:
Add ToDo class

woojiahao committed:
Add Deadline class

woojiahao committed:
Add Event class

woojiahao committed:
Refactor to use switch statement

woojiahao committed:
Use setDone instead of marking in Task

woojiahao committed:
Add addToDo method

woojiahao committed:
Add deadline handling

woojiahao committed:
Add event creation handler

woojiahao committed:
Clean up command handling

woojiahao committed:
Setup semi-automated test suite

woojiahao committed:
Handle command not found

woojiahao committed:
Use vararg for printText

woojiahao committed:
Move all tasks to separate package

woojiahao committed:
Remove unused import

woojiahao committed:
Handle empty input

woojiahao committed:
Use list of strings for addToDo

woojiahao committed:
Refactor deadline creation

woojiahao committed:
Refactor event creation

woojiahao committed:
Clean up mark/unmark

woojiahao committed:
Clean up output

woojiahao committed:
Expand test suite

woojiahao committed:
Add error for /to before /from

woojiahao committed:
Trim all inputs

woojiahao committed:
Create TaskList singleton to centralise tasks

woojiahao committed:
Create PrintUtility to print formatted output

woojiahao committed:
Add ICommand to centralise command dispatch

woojiahao committed:
Add AddTodo command

woojiahao committed:
Add AddEvent command

woojiahao committed:
Add AddDeadline event

woojiahao committed:
Add ListTasks command to print task list

woojiahao committed:
Add DeleteTask command

woojiahao committed:
Add Mark/UnmarkTask commands to change task status

woojiahao committed:
Add Unknown command to handle invalid commands

woojiahao committed:
Use command pattern to dispatch behavior

woojiahao committed:
Flatten input and apply indentation to all lines

woojiahao committed:
Move list tasks behavior to TaskList

woojiahao committed:
Add data file parsing from JSON

woojiahao committed:
Setup save tasks after action

woojiahao committed:
Use GSON annotations to serialize Tasks

woojiahao committed:
Use try-with-resources for loading tasks

woojiahao committed:
Add serialized type field for deadline

woojiahao committed:
Merge branch 'branch-level-7'

woojiahao committed:
Parse date string as LocalDate for add deadline

woojiahao committed:
Handle dates for event task type

woojiahao committed:
Merge branch 'branch-Level-8'

woojiahao committed:
Use more OOP + Dependency Injection

woojiahao committed:
Merge branch 'A-MoreOOP'

woojiahao committed:
Use object equals for ParseInfo.EMPTY

woojiahao committed:
Move to cyrus/ folder

woojiahao committed:
Merge branch 'A-Packages'

woojiahao committed:
Merge remote-tracking branch 'origin/add-gradle-support'

woojiahao committed:
Add Gradle dependency

woojiahao committed:
Fix parser bug with empty input

woojiahao committed:
Add test for empty input

woojiahao committed:
Add unit tests for parse input

woojiahao committed:
Add test for option parsing

woojiahao committed:
Fix option parsing

Primary issue was that options did not have any whitespace and lone /
would also cause the parser to treat it as an empty key

woojiahao committed:
Add test for parsing without options

woojiahao committed:
Add unit test for CommandType parsing

woojiahao committed:
Setup FileStorage tests

woojiahao committed:
Merge branch 'A-JUnit'

woojiahao committed:
Fix main class path

woojiahao committed:
Implement find command

woojiahao committed:
Add documentation

woojiahao committed:
Add more JavaDocs

woojiahao committed:
Clean up documentation

woojiahao committed:
Run Intellij's formatter once

woojiahao committed:
Use full uppercase for constants

woojiahao committed:
Abide by coding style

woojiahao committed:
Merge branch 'branch-Level-9'

woojiahao committed:
Merge branch 'branch-A-JavaDoc'

woojiahao committed:
Add JavaDocs for FindTaskCommand

woojiahao committed:
Merge branch 'branch-A-CodingStandard'

woojiahao committed:
Format tests

woojiahao committed:
Fix bugs from merge conflicts

woojiahao committed:
Fix runtest script to work with Gradle

woojiahao committed:
Fix spelling issues

woojiahao committed:
Setup Checkstyle

woojiahao committed:
Fix all checkstyle issues

woojiahao committed:
Merge branch 'branch-A-CheckStyle'

woojiahao committed:
Add misc documentation

woojiahao committed:
Fix jar filename in runtest

woojiahao committed:
Setup JavaFX dependencies

woojiahao committed:
Rename current Ui to CliUi

Ensures backward compatibility while migrating to use JavaFX

woojiahao committed:
Setup base JavaFX project

woojiahao committed:
Setup initial JavaFX styling

woojiahao committed:
Setup echo user input

woojiahao committed:
Setup individual DialogBox for GUI

woojiahao committed:
Add support for alignment text

woojiahao committed:
Refactor command system to return String[]

Parse the String[] and display in the GUI

woojiahao committed:
Style Cyrus GUI

woojiahao committed:
Merge branch 'branch-Level-10'

jordankanghm commented on src/main/java/cyrus/Cyrus.java:
```diff
@@ -0,0 +1,51 @@
+     * @return {@code String} output of command if successfully executed.
+     * @throws CommandError if there is an error with the input.
+     */
+    public String dispatchAndExecute(ParseInfo parseInfo) throws CommandError {
```
Good choice of name which clearly states the purpose of the method!

jordankanghm commented on src/main/java/cyrus/adapters/LocalDateAdapter.java:
```diff
@@ -0,0 +1,42 @@
+ */
+public class LocalDateAdapter extends TypeAdapter<LocalDate> {
+    /**
+     * When writing {@code LocalDateAdapter} to JSON, convert it to the input format of {@code dd
```
Perhaps starting with a continuous verb could better adhere to JavaDoc conventions? E.g. "Converts the code from..."

jordankanghm commented on src/main/java/cyrus/commands/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,54 @@
+     */
+    @Override
+    public String[] execute() throws CommandError {
+        if (this.parseInfo.hasNoArgument()) {
```
Great segmenting of logic via blank lines!

jordankanghm commented on src/main/java/cyrus/commands/CommandType.java:
```diff
@@ -0,0 +1,48 @@
+     * @return mapped {@code CommandType}
+     */
+    public static CommandType fromString(String input) {
+        switch (input) {
```
Great job implementing the correct coding conventions for switch case!

jordankanghm commented on src/main/java/cyrus/parser/ParseInfo.java:
```diff
@@ -0,0 +1,75 @@
+ * <p>Modelled after CLI applications where the {@code argument} is the text before any options
+ * and an {@code option} is the values of "flags".</p>
+ */
+public class ParseInfo {
```
Perhaps the class name could be a noun instead of a verb?

geraldngjx replied:
I like how you have implemented this class to parse the commands more efficiently! However, perhaps you can consider doing this within Parser class instead?

jordankanghm commented on src/main/java/cyrus/tasks/Task.java:
```diff
@@ -0,0 +1,40 @@
+public class Task {
+    public final String name;
+    @SerializedName("status")
+    private boolean done;
```
Perhaps the naming could sound more like a boolean, maybe "isDone"?

jordankanghm reviewed (commented):
Overall the segmenting of logic in methods are great! One improvement could be the first line of JavaDocs. Try to start the line with a verb ending with "s"!

geraldngjx commented on src/main/java/cyrus/Cyrus.java:
```diff
@@ -0,0 +1,51 @@
+     * @return parsed user input.
+     */
+    public ParseInfo parseInput(String input) {
+        return parser.parse(input);
```
I like how you have named your method as it is clear and descriptive!

geraldngjx commented on src/main/java/cyrus/Cyrus.java:
```diff
@@ -0,0 +1,51 @@
+     * @param input user input to parse.
+     * @return parsed user input.
+     */
+    public ParseInfo parseInput(String input) {
```
I like how you have named your method as it is clear and descriptive!

geraldngjx commented on src/main/java/cyrus/storage/FileStorage.java:
```diff
@@ -0,0 +1,162 @@
+                    .create();
+    private final String dataFilePath;
+
+    public FileStorage(String dataFilePath) {
```
Perhaps you can consider adding JavaDocs for the constructor as well?

geraldngjx commented on src/main/java/cyrus/utility/DateUtility.java:
```diff
@@ -0,0 +1,48 @@
+/**
+ * Utility to deal with {@code LocalDate}.
+ */
+public class DateUtility {
```
I like how you have abstracted most of your code into distinct objects that is easy to read and understand!

geraldngjx reviewed (commented):
Great job with the naming conventions and adherence to coding standards! 👍 Your variable names are clear and descriptive, which makes the code easy to read and understand. Consistent coding style throughout the codebase is crucial for maintainability, and you've done an excellent job in that regard. Keep up the good work!

geraldngjx commented on src/main/java/cyrus/parser/Parser.java:
```diff
@@ -0,0 +1,101 @@
+     * @return appropriate {@code Command}
+     */
+    public Command dispatchCommand(TaskList taskList, ParseInfo parseInfo) {
+        switch (parseInfo.getCommandType()) {
```
Good use of switch to keep your code readable and clean!

geraldngjx reviewed (commented):
Overall, great job in implementing clean and readable code! I left a few more comments and you might consider perhaps handling certain task (e.g. ParseInfo) within existing classes too. Excellent work!

SungMatt commented on README.md:
````diff
@@ -1,24 +1,46 @@
-   | |_| | |_| |   <  __/
-   |____/ \__,_|_|\_\___|
-   ```
+# Cyrus
+
+> Worry no more about your tasks!
````
Nicely done with the readme!

SungMatt commented on src/main/java/cyrus/storage/FileStorage.java:
```diff
@@ -0,0 +1,162 @@
+    private void enforceFields(HashMap<String, String> map) {
+        String[] mandatoryKeys = {"type", "status", "name"};
+        Consumer<String[]> checkKeys = (keys) -> {
+            for (String key : keys) {
```
Nicely done with the new style of for statement! Good work!

SungMatt commented on src/main/java/cyrus/tasks/TaskList.java:
```diff
@@ -0,0 +1,108 @@
+        return this.tasks
+                .stream()
+                .filter(task -> task.name.contains(keyword))
+                .collect(Collectors.toList());
```
Good use of streams!

SungMatt commented on src/test/java/cyrus/storage/FileStorageTest.java:
```diff
@@ -0,0 +1,93 @@
+
+public class FileStorageTest {
+    @Test
+    public void testLoadFileCreation() {
```
nit -- but the three part naming convention of tests is useful here! What's the expected behavior?

SungMatt reviewed (approved):
Great work, Jia Hao! Love the work you have done! LGTM!

woojiahao committed:
Use assert to enforce FileStorage assumptions

FileStorage currently does not enforce the dataFilePath and has a
slightly awkward way of handling missing keys in the parsed data.

This runs the risk of having invalid input and the use of
IllegalStateException does not make a lot of sense in the long run.

Adopt assert statements to ensure dataFilePath is not empty and
AssertionError is thrown instead of IllegalStateException for missing
keys.

This improves the visibility of configuration issues and makes it
clearer why exactly the bot failed.

woojiahao committed:
Add more unit tests for FileStorage invalid cases

The unit tests for FileStorage currently does not verify empty file
names and invalid data formats.

This creates missing test coverage that could lead to bugs.

Add two separate unit tests to ensure that empty file names and invalid
data formats both properly throw AssertionErrors.

This ensures the behavior are covered.

woojiahao committed:
Improve the JavaDocs of LocalDateAdapter

The JavaDocs of LocalDateAdapter currently is worded ambiguously.

This creates confusion when reading them.

Reworded the documentation to be clearer.

woojiahao committed:
Create "magic" variable for option lead

woojiahao committed:
Abstract valid part checking to method

woojiahao committed:
Remove public keyword for interface methods

woojiahao committed:
Rename user and bot image variables

woojiahao committed:
Add text color option for GUI

woojiahao committed:
Merge pull request #1 from woojiahao/branch-A-Assertions

Add usage of `assert` keyword

woojiahao committed:
Merge branch 'master' into branch-A-CodeQuality

woojiahao committed:
Merge pull request #2 from woojiahao/branch-A-CodeQuality

Implement code quality changes

woojiahao committed:
Add unit test CI action for Github

woojiahao committed:
Fix wildcard import

woojiahao committed:
Merge pull request #3 from woojiahao/branch-A-CI

Add unit test CI action for Github

woojiahao committed:
Setup skeleton command to handle statistics

woojiahao committed:
Add command return behavior for viewing statistics

woojiahao committed:
Add pie chart to visualise task distribution

woojiahao committed:
Add completedDate and separate mark/unmark method

woojiahao committed:
Parse completedDate field from JSON

woojiahao committed:
Render weekly task completion rate distribution

woojiahao committed:
Fix coding standard violation

woojiahao committed:
Merge pull request #4 from woojiahao/branch-C-Statistics

Add statistics dashboard

woojiahao committed:
Use IllegalStateException as assert may be ignored

woojiahao committed:
Reduce load method length

woojiahao committed:
Abstract data loading to separate methods

woojiahao committed:
Address JavaDoc comments

woojiahao committed:
Apply base style tweaks

woojiahao committed:
Add welcome text on chat start

woojiahao committed:
Fix broken unit test

woojiahao committed:
Merge pull request #6 from woojiahao/branch-A-BetterGui

Add better GUI

woojiahao committed:
Add initial user guide

woojiahao committed:
Add initial user guide

woojiahao committed:
Add details for keywords

woojiahao committed:
Merge branch 'master' into branch-A-UserGuide

woojiahao committed:
Merge pull request #7 from woojiahao/branch-A-UserGuide

Add user guide

woojiahao committed:
Add Ui.png

woojiahao committed:
Add name to title and update Ui.png

woojiahao committed:
Add unit testing for DateUtility

woojiahao committed:
Add unit test for AddDeadlineCommand

woojiahao committed:
Add unit test for AddEventCommandTest

woojiahao committed:
Add unit test for AddTodoCommand

woojiahao committed:
Fix delete task output spelling

woojiahao committed:
Throw CommandError if no find term present

woojiahao committed:
Add unit test for ByeCommand

woojiahao committed:
Add unit test for DeleteTaskCommand

woojiahao committed:
Add unit test for FindTaskCommand

woojiahao committed:
Add unit tests for ListTasksCommand

woojiahao committed:
Add unit tests for MarkTaskCommand

woojiahao committed:
Add unit tests for UnknownCommand

woojiahao committed:
Add unit tests for UnmarkTaskCommand

woojiahao committed:
Add unit tests for ViewStatisticsCommand

woojiahao committed:
Clean up TaskList code structure

woojiahao committed:
Expand all wildcard imports

woojiahao committed:
Clean up error handling logic for FileStorage

woojiahao committed:
Clean up JavaDocs

woojiahao committed:
Resolve cross-platform issue

woojiahao committed:
Clean up MainWindow code

woojiahao committed:
Fix TaskList JavaDocs

woojiahao committed:
Reorganize JavaFX components in folder

woojiahao committed:
Clean up documentation

woojiahao committed:
Fix usage of mark, unmark, and delete

woojiahao committed:
Explain what happens when no tasks for statistics
